python/pyspark/sql/connect/streaming/worker/foreach_batch_worker.py

```diff
@@ -69,8 +73,32 @@ def process(df_id, batch_id):  # type: ignore[no-untyped-def]
     while True:
         df_ref_id = utf8_deserializer.loads(infile)
         batch_id = read_long(infile)
-        process(df_ref_id, int(batch_id))  # TODO(SPARK-44463): Propagate error to the user.
-        write_int(0, outfile)
+        # Handle errors inside Python worker. Write 0 to outfile if no errors and write -2 with
+        # traceback string if error occurs.
+        try:
+            process(df_ref_id, int(batch_id))
+            write_int(0, outfile)
+        except BaseException as e:
```

Comment:
I think we should probably put this handling into a util and reuse it in the regular `worker.py` too.

Reply:
you could put that into `pyspark.util`

Reply:
done.